torchinductor/triton_ops/autotune.py

```diff
@@ -0,0 +1,170 @@
+    if z:
+        z = min(z, size_hints[2])
+
+    # if we are below original block size, scale up where we can
+    while x < size_hints[0] and conditional_product(x, y, z) < target:
+        x *= 2
+    while y and y < size_hints[1] and conditional_product(x, y, z) < target:
+        y *= 2
+    while z and z < size_hints[2] and conditional_product(x, y, z) < target:
+        z *= 2
+
+    cfg = {"XBLOCK": x}
+    if y:
+        cfg["YBLOCK"] = y
+    if z:
+        cfg["ZBLOCK"] = z
+    return Config(cfg, num_warps=num_warps, num_stages=num_stages)
+
+
+def triton_config_reduction(size_hints, x, r, num_warps=4, num_stages=2):
+    """
+    Construct a reduction triton config with some adjustment heuristics
+    based on size_hints. Size_hints is a tuple of numels in each tile
+    dimension and will be rounded up to the nearest power of 2.
+    """
+
+    target = conditional_product(x, r)
+    if conditional_product(*size_hints) < target:
+        target //= 8
+
+    # shrink sizes to size hints
+    x = min(x, size_hints[0])
+    r = min(r, size_hints[1])
+
+    # if we are below original block size, scale up where we can
+    while x < size_hints[0] and conditional_product(x, r) < target:
+        x *= 2
+    while r < size_hints[1] and conditional_product(x, r) < target:
+        r *= 2
+
+    cfg = {"XBLOCK": x, "RBLOCK": r}
+    return Config(cfg, num_warps=num_warps, num_stages=num_stages)
+
+
+def apply_triton_config(config):
+    """
+    Decorator that applies a fixed triton config using triton.heuristics.
+    """
+
+    def getter(name):
+        def get(args):
+            return config.kwargs[name]
+
+        return get
+
+    return heuristics({name: getter(name) for name in config.kwargs.keys()})
+
+
+def pointwise_heuristics(size_hints):
+    """
+    Construct @triton.heuristics() based on size_hints.
+    """
+
+    if len(size_hints) == 1:
+        return apply_triton_config(triton_config(size_hints, 1024))
+    if len(size_hints) == 2:
+        if not config.triton.autotune:
+            return apply_triton_config(triton_config(size_hints, 64, 64))
+        return autotune(
+            [
+                # TODO(jansel): do warps matter?
+                triton_config(size_hints, 64, 64),
+                triton_config(size_hints, 4, 256),
+                triton_config(size_hints, 256, 4),
+            ],
+            key=["xnumel", "ynumel"],
+        )
+    if len(size_hints) == 3:
+        return apply_triton_config(triton_config(size_hints, 16, 16, 16))
+    raise NotImplementedError(f"size_hints: {size_hints}")
+
+
+def reduction_heuristics(size_hints):
+    """args to @triton.heuristics()"""
+
+    if len(size_hints) == 2:
+        if not config.triton.autotune:
+            return apply_triton_config(triton_config_reduction(size_hints, 32, 128))
+        return autotune(
+            [
+                # TODO(jansel): do warps matter?
+                triton_config_reduction(size_hints, 32, 128, num_stages=3),
```

Comment:
I don't think `num_stages` matter, they are more for mm code, `warps` probably matter, for larger reduction dimensions especially when reduction is followed by broadcast you want more warps, see e.g. in `layer_norm` tutorial `num_warps = min(max(BLOCK_SIZE // 256, 1), 8)`, `BLOCK_SIZE` is capped reduction size there

Reply:
I'll try this in a followup PR